pyccl/halos/halo_model.py

```diff
@@ -23,83 +23,69 @@ class HMCalculator(CCLAutoreprObject):
-                 k_min=1E-5):
-        # halo mass definition
-        if isinstance(mass_def, MassDef):
-            self._mdef = mass_def
-        elif isinstance(mass_def, str):
-            self._mdef = MassDef.from_name(mass_def)()
-        else:
-            raise TypeError("mass_def must be of type `MassDef` "
-                            "or a mass definition name string")
-
-        # halo mass function
-        if isinstance(massfunc, MassFunc):
-            self._massfunc = massfunc
-        elif isinstance(massfunc, str):
-            nMclass = MassFunc.from_name(massfunc)
-            self._massfunc = nMclass(cosmo, mass_def=self._mdef)
-        else:
-            raise TypeError("mass_function must be of type `MassFunc` "
-                            "or a mass function name string")
-
-        # halo bias function
-        if isinstance(hbias, HaloBias):
-            self._hbias = hbias
-        elif isinstance(hbias, str):
-            bMclass = HaloBias.from_name(hbias)
-            self._hbias = bMclass(cosmo, mass_def=self._mdef)
-        else:
-            raise TypeError("halo_bias must be of type `HaloBias` "
-                            "or a halo bias name string")
-
-        self._prec = {'log10M_min': log10M_min,
-                      'log10M_max': log10M_max,
-                      'nlog10M': nlog10M,
-                      'integration_method_M': integration_method_M,
-                      'k_min': k_min}
-        self._lmass = np.linspace(self._prec['log10M_min'],
-                                  self._prec['log10M_max'],
-                                  self._prec['nlog10M'])
+    __repr_attrs__ = ("mass_function", "halo_bias", "mass_def", "precision",)
+    __getattr__ = deprecate_attr(pairs=[('_mdef', 'mass_def'),
+                                        ('_mdef', 'mass_def'),
+                                        ('_massfunc', 'mass_function'),
+                                        ('_hbias', 'halo_bias'),
+                                        ('_prec', 'precision')]
+                                 )(super.__getattribute__)
+
+    @warn_api(pairs=[("massfunc", "mass_function"), ("hbias", "halo_bias"),
+                     ("log10M_min", "lM_min"), ("log10M_max", "lM_max"),
+                     ("nlog10M", "nlM"), ("k_min", "k_norm")])
+    def __init__(self, *, mass_function, halo_bias, mass_def=None,
+                 lM_min=8., lM_max=16., nlM=128,
+                 integration_method_M='simpson', k_norm=1E-5):
+        # Initialize halo model ingredients
+        self.mass_def = MassDef.initialize_from_input(mass_def)
+        kw = {"mass_def": self.mass_def}
+        self.mass_function = MassFunc.initialize_from_input(mass_function, **kw)  # noqa
+        self.halo_bias = HaloBias.initialize_from_input(halo_bias, **kw)
+
+        # Check mass definition consistency.
+        if not (self.mass_def
+                == self.mass_function.mass_def
+                == self.halo_bias.mass_def):
```

Comment:
Is this using the `_repr`? if so, implement an `__eq__`

Reply:
This is using the `_repr` here but after #1033 which is awaiting review it will automatically switch to `__eq__`.

Reply:
Another question I'd like to ask you is - do we need this check? Or should we allow users to pass in whatever they want? Mass functions, halo biases etc. are only accurate to ~O(10%), so idk if we should do this or not.

Reply:
yes, this check is vital. Please implement __eq__ for massdefs in this PR, I do not want to put anything in the code that jeopardises efficiency. who knows when or if the other PR will be merged.

Reply:
done